### DISCUSSION TITLE
SendGrid Webhook

### DISCUSSION THREAD

**Discussion by imnotberg** [anymail/django-anymail#215] [Q&A]
I am using the sendgrid module of this program and I have set up my webhooks url with the Sendgrid settings.

However, with any sendgrid event that I use, the only post I get is that I've had a SendGrid API Event.  No real metadata beyond that.

How do I understand more about the actual event rather than just that there was an event?

**Comment by medmunds** [comment]
@imnotberg sorry, I'm having a little trouble understanding your question.

If you're asking about how to interpret the `AnymailTrackingEvent` that's passed to your webhook signal handler function, that's documented under [*Tracking sent mail status*](https://anymail.readthedocs.io/en/stable/sending/tracking/). Here's an example, but see the docs for more properties and information:

```python
from anymail.signals import tracking
from django.dispatch import receiver

@receiver(tracking)
def handle_tracking_event(sender, event, esp_name, **kwargs):
    print(event.event_type)  # 'bounced' or 'delivered' or 'opened' or ....
    print(event.recipient)   # e.g., 'recipient@example.com'
    print(event.message_id)  # matches message.anymail_status.message_id from sending
    # etc.
```

If you already found that documentation and set up your signal handler, and your question is that the event properties don't seem to be set to useful values, it would help to know: (1) your signal handler code, (2) the actual event values you *are* getting, if any, in that handler, and (3) what your SendGrid Event Webhook [Mail Settings](https://app.sendgrid.com/settings/mail_settings) are—particularly which of the "Events to be POSTed to your URL" you've enabled.

Or if it's neither of those, please provide more details about what you're trying to do, how you're trying to do it (showing your code is usually helpful), what you expected to happen and what's actually happening instead.

**Reply by imnotberg** [reply]
Thanks for getting back to me so quickly.  

I think my question is a little more foundational than that.  I see the webhook working on my logs, but I can't figure out how to do anything with the info being passed. 

Where should this receiver live?  Should it be on the view that's being processed? I just can't seem to do anything with the information but I can see that sendgrid is definitely posting to my app.

[18/Dec/2020:17:41:49 +0000] "POST /sendgrid_webhook/ HTTP/1.1" 200 0 "-" "SendGrid Event API" -- but that's all that the post says is "EVENT API" really cant' do much with that.

thanks for any help.

**Reply by medmunds** [reply]
Ah, that helps clarify what you're asking, thanks.

First, in your Django server logs…
> [18/Dec/2020:17:41:49 +0000] "POST /sendgrid_webhook/ HTTP/1.1" 200 0 "-" "SendGrid Event API"

… "SendGrid Event API" is the [user agent](https://en.wikipedia.org/wiki/User_agent) that is posting to your Django server. (For web pages real users are visiting, you'll see long strings including browser names like "Safari" or "Chrome" in that position.) But that's not the actual content of the post—server logs typically don't include the actual content for *any* post, because that would leak all kinds of sensitive information into the logs (like people's passwords on your login page).

In a normal Django view, you write code (or use Django forms) that examines `request.POST` or `request.body` to get at the content of the post. And if you wanted, you could do that for a SendGrid webhook too—just write your own view function, give that URL to SendGrid, and leave Anymail out of the picture.

When you use Anymail's webhook urls, Anymail has written those view functions for you. Anymail's webhook views parse the data from SendGrid (and other ESPs) into a standardized `AnymailTrackingEvent` object, and then notify your code using the [Django signals](https://docs.djangoproject.com/en/stable/topics/signals/) mechanism.

To receive those notifications, you need to write and "connect" one or more "signal receiver" functions that "[listen to](https://docs.djangoproject.com/en/stable/topics/signals/#listening-to-signals)" Anymail's custom `tracking` signal.

> Where should this receiver live?

Excellent question. The Django docs have some guidance: see the green box titled "Where should this code live?" under [*Connecting receiver functions*](https://docs.djangoproject.com/en/stable/topics/signals/#connecting-receiver-functions). But if you just want to see it working, you can stick the function anywhere that will get imported as your app starts up. Say, in your urls.py or views.py. (But put it at the root of the file, not nested as a method inside some class—the `@receiver` that registers it needs to run as the file is imported. Also, for long term maintainability, you'll eventually want to move the receiver function to a separate signals.py file that you load from your [`AppConfig.ready()`](https://docs.djangoproject.com/en/stable/ref/applications/#django.apps.AppConfig.ready) as Django recommends.)

Once you've got your tracking webhook signal receiver connected, if you put `print(event.___)` statements in it like the earlier example, you should start seeing all sorts of interesting tracking event data.

Hope that helps. Let me know if you run into any trouble with it.
